### PR TITLE
8.2.2_update_teachers_profile_kozhukh

### PR DESCRIPTION
**Pull request by OksanaKozhukh**

### CODE REVIEW

File: src/modules/teachers/teachers.service.ts [hunk 1]
```diff
@@ -1,19 +1,42 @@
...
 
     return teacher;
   }
+
+  public async updateOne(id: number, data: PartialUpdate<ITeachersData>, user: Teachers) {
+    if (id !== user.id && !user.isAdmin) {
+      throw new Unauthorized('You cannot change another profile');
+    }
+
+    return await sequelize.transaction(async (transaction) => {
```
Comment (Alan18081) [inline src/modules/teachers/teachers.service.ts:34]: It is one database query, we don't need actual transaction here

File: src/common/types/types.ts [hunk 2]
```diff
@@ -4,3 +4,8 @@ import {Students} from '../../modules/students/students.model';
 
 export type CustomUser = (Teachers  | Students) & {isMentor: boolean; groupId?: number; isAdmin?: boolean};
 export type AuthRequest = Request & {user: CustomUser};
+
+export type UpdateRequest<T>= Request & {user: T};
+export type PartialUpdate<T> = {
```
Comment (Alan18081) [inline src/common/types/types.ts:9]: Typescript already has implemented the `Partial` mapped type, please use it. But It is great that you have spent time understanding how mapping types work. Great job, keep going

File: src/modules/teachers/teachers.controller.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { Request, Response, NextFunction } from 'express';
+import { teachersService } from './teachers.service';
+import { Teachers } from './teachers.model';
+import { UpdateRequest } from '../../common/types/types';
+import { validateIdOrThrow } from '../../common/validators/';
+
+class TeachersController {
+    public async updateOne(req: UpdateRequest<Teachers>, res: Response,
+                           next: NextFunction): Promise<void> {
+        validateIdOrThrow(+req.params.id);
```
Comment (Alan18081) [inline src/modules/teachers/teachers.controller.ts:10]: Put it inside `try...catch`, please

File: src/modules/teachers/teachers.service.ts (hunk 1 above)
Comment (OksanaKozhukh) [reply src/modules/teachers/teachers.service.ts]: done

File: src/common/types/types.ts (hunk 2 above)
Comment (OksanaKozhukh) [reply src/common/types/types.ts]: done

File: src/modules/teachers/teachers.controller.ts (hunk 3 above)
Comment (OksanaKozhukh) [reply src/modules/teachers/teachers.controller.ts]: done

File: src/modules/teachers/teachers.controller.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { Request, Response, NextFunction } from 'express';
+import { teachersService } from './teachers.service';
+import { Teachers } from './teachers.model';
+import { UpdateRequest } from '../../common/types/types';
+import { validateIdOrThrow } from '../../common/validators/';
+
+class TeachersController {
+    public async updateOne(req: UpdateRequest<Teachers>, res: Response,
```
Comment (antonpanchenk0) [inline src/modules/teachers/teachers.controller.ts:8]: why you use `UpdateRequest<Teachers>`? Why not use `AuthRequest`

Comment (OksanaKozhukh) [reply src/modules/teachers/teachers.controller.ts:8]: Alex recommended me to use generic type and split it into 2 separate parts for Students and Teachers

File: src/modules/teachers/teachers.service.ts [hunk 5]
```diff
@@ -21,6 +31,18 @@ class TeachersService {
     return teacher;
   }
 
+
+  public async updateOne(id: number, data: Partial<ITeachersData>, user: Teachers) {
+    if (id !== user.id && !user.isAdmin) {
+      throw new Unauthorized('You cannot change another profile');
+    }
+
```
Comment (Alan18081) [inline src/modules/teachers/teachers.service.ts:39]: Please, add here validation that this mentor exists in database if user is super admin

Comment (OksanaKozhukh) [reply src/modules/teachers/teachers.service.ts:39]: done

File: src/modules/teachers/teachers.service.ts [hunk 6]
```diff
@@ -23,7 +20,7 @@ class TeachersService {
     return teacher;
   }
 
-  public async findOneById(id: number) {
+  public async findOneById(id: number, transaction: Transaction) {
     const teacher = await Teachers.findOne({
       where: { id },
```
Comment (Alan18081) [inline src/modules/teachers/teachers.service.ts:25]: Please, use the transaction as a part of request to dabatabase

Comment (OksanaKozhukh) [reply src/modules/teachers/teachers.service.ts]: done

**Pull request merged**